torch/fx/passes/reinplace.py

```diff
@@ -0,0 +1,257 @@
+import torch
+import torch.fx as fx
+from torch.fx._compatibility import compatibility
+
+from torch.multiprocessing.reductions import StorageWeakRef
+
+from collections import defaultdict
+from typing import Iterable
+import itertools
+
+__all__ = ['reinplace']
+
+def concatMap(func, xs):
+    for x in xs:
+        for r in func(x):
+            yield r
+
+class GatherReinplaceDatastructures(torch.fx.Interpreter):
+    def __init__(self, gm):
+        super().__init__(gm)
+        self.node_counter = -1
+        # graph variable names of all program inputs
+        self.input_names = []
+        # Groups sets of the variable names of all aliased variables in the program
+        self.alias_map = defaultdict(set)
+        # For a given variable, tells you the names of all nodes in the program that use it.
+        # (including the index of that node, so we know where in the program it is)
+        self.names_to_used_nodes = defaultdict(list)
+        # When you call `x.copy_(y)`, you overwrite x's data.
+        # This stores the name of the node (if any) where a given variable in the program is overwritten.
+        # We need to track this information for program inputs
+        self.last_node_where_variable_is_overwritten = defaultdict(lambda: -1)
```

Comment:
So if we setup the tiered representation, where `copy_` is kept out of the IR, then we actually wouldn't have to deal with this; instead, we would have a separate optimization pass which "collapses" the tiered rep.

Actually this makes me wonder if we should represent the tiering as just a nested module. So you have an inner GraphModule which is purely functional, and then the outer container GraphModule which has the copies. The alternative was to have an extra epilogue module but I wonder if explicitly having the nesting makes it clearer. One downside is that all optimization passes would need to know to "look past" the containing module into the functional inside, so maybe this is not so great.

Reply:
Another possibility is we create a GraphModule subclass where the graph is functional, but it has some extra metadata saying what `copy_` to do and we override the codegen to insert these copies when you actually generate the code for the module. 

I wonder if @Chillee has an opinion.

Reply:
It seems to me like in a lot of cases, re-inplacing will basically turn the side-car `copy_()` module/graph into a no-op (by folding it back into a reinplace on the input). I guess the main case where this wouldn't happen is if we're using a fuser and the fuser already added the op into a fusion group.

I don't have a strong opinion, but it feels easy / clean just to keep the `copy_()`'s in a separate module, and leave it up the re-inplacing to optionally fold them back into the graph:
```
# output represents the newly inplaced module, and the new version of the `copy_()` module (where some of those copies may have been removed, and folded into the first module).
def reinplace(functional_graph: fx.GraphModule, input_mutation_graph: fx.GraphModule) -> Tuple[fx.GraphModule, fx.GraphModule]
```